jupyterhub/scopes.py

```diff
@@ -569,6 +561,71 @@ def expand_scopes(scopes, owner=None, oauth_client=None):
     return frozenset(reduce_scopes(expanded_scopes))
 
 
+def _resolve_requested_scopes(requested_scopes, have_scopes, user, client, db):
+    """Resolve requested scopes for an OAuth token
+
+    Intersects requested scopes with user scopes.
+
+    First, at the raw scope level,
+    then if some scopes remain, intersect expanded scopes.
+
+    Args:
+        requested_scopes (set):
+            raw scopes being requested.
+        user (orm.User):
+            user for whom the scopes will be issued
+        client (orm.OAuthClient):
+            oauth client which will own the token
+
+    Returns:
+        (allowed_scopes, excluded_scopes): sets of allowed scopes and excluded scopes from the request
+    """
+
+    allowed_scopes = requested_scopes.intersection(have_scopes)
+    excluded_scopes = requested_scopes.difference(have_scopes)
```

Comment:
There is a mix of terminology here. `allowed` / `disallowed` is used outside this function, while inside it `allowed` and `excluded` is used.

I appreciate the `allowed` and `disallowed` overall, and currently lean towards thinking it would be an improvement to update the function naming to speak of `disallowed_scopes`.

I'm open to retaining naming as it is though if you think it makes sense.